design/fast_finality.md

```diff
@@ -0,0 +1,190 @@
+* How do we ensure the sequencer distributes all light batches to all nodes?
+* Do the light batches need to be linked to the latest block that was fed into the enclave?
+* How do we achieve high-availability of the sequencer, both for the enclave and for the host? How do we prevent 
+  denial-of-service attacks on both?
+* How do we prevent attackers from "gumming up" the sequencer with random, fake light-batches?
+
+## Appendices
+
+### Designs considering for preventing value-extraction
+
+In this section, we investigate various designs to prevent value-extraction. The specific attack we have in mind is one 
+where the sequencer runs `n` enclaves, and uses `n-1` of them to test the impact of various transaction sets to 
+identify value-extraction opportunities.
+
+#### Do nothing
+
+In this approach, we rely on trust in the sequencer and the fact that value-extraction opportunities are reduced in
+Obscuro due to its data-visibility rules.
+
+#### Disable `eth_call` on sequencer enclaves
+
+By disabling `eth_call` on the sequencer enclaves, we prevent the operator from extracting information about the impact
+of a given light batch.
+
+This approach is unworkable. The operator can run a separate, non-sequencer enclave, feed the light batches to that
+enclave, and use `eth_call`s on that enclave to determine the impact of a given light batch.
+
+#### Run a single sequencer enclave
+
+If there is a single sequencer enclave, there are no other enclaves to use to identify value-extraction opportunities. 
+The single sequencer cannot be restarted to identify value-extraction opportunities, since the enclave start-up delay 
+will then prevent the sequencer from reaching its block production target.
+
+This approach is unworkable. We cannot achieve the desired high-availability with a single sequencer enclave.
+
+#### Detect restarts on sequencer enclaves
+
+We allow validators to detect how often sequencer enclaves are restarted, incentivising good behaviour on behalf of the 
+sequencer, and allowing the issue of a malicious (or incompetent) sequencer to be detected and handled as a governance 
+action. For this to work, validators would have to actively assess whether the sequencer is doing an adequate job.
+
+There are two flavours of this.
+
+##### Have sequencer enclaves produce lifetime proofs
+
+Every `x`th light batch includes a proof of how long (e.g. in terms of light batches or L1 blocks) each sequencer 
+enclave has been up. This creates a history of when each sequencer enclave was restarted. `x` can be arbitrarily high, 
+since you can work backwards from this proof and the previous proof to check the enclave has been up the entire time.
+
+An alternative model is for each sequencer enclave to post a restart proof to the L1 whenever it starts, and wait for 
+that proof to be included in the L1 before producing light batches and rollups (thus forcing the proof to be posted on 
```

Comment:
It doesn't need to be L1 necessarily. Just requires some kind of ack that it's been published (so inclusion in an LB would work I guess)

Reply:
if it doesn't post to the L1, what stops the host hiding the proof from other nodes?

Reply:
Yeah that's the "some kind of ack" part. It being published in an LB would be enough though I think.

Reply:
Fixed, thanks